hoomd/md/NeighborList.cc

```diff
@@ -1742,6 +1769,252 @@ void NeighborList::updateMemoryMapping()
     }
 #endif
 
+pybind11::array_t<uint32_t> NeighborList::getLocalPairListPython(uint64_t timestep)
+    {
+    compute(timestep);
+
+    bool third_law = getStorageMode() == NeighborList::half;
+
+    ArrayHandle<unsigned int> h_n_neigh(getNNeighArray(), access_location::host, access_mode::read);
+    ArrayHandle<unsigned int> h_nlist(getNListArray(), access_location::host, access_mode::read);
+    ArrayHandle<size_t> h_head_list(getHeadList(), access_location::host, access_mode::read);
+
+    auto* pair_list = new std::vector<vec2<uint32_t>>();
+
+    pybind11::capsule delete_when_done(pair_list,
+                                       [](void* f)
+                                       {
+                                           auto data
+                                               = reinterpret_cast<std::vector<vec2<uint32_t>>*>(f);
+                                           delete data;
+                                       });
+
+    for (unsigned int i = 0; i < m_pdata->getN(); i++)
+        {
+        const size_t my_head = h_head_list.data[i];
+        const unsigned int size = h_n_neigh.data[i];
+        for (unsigned int k = 0; k < size; k++)
+            {
+            const unsigned int j = h_nlist.data[my_head + k];
+            // if j is not a ghost, only accept it if i < j
+            if (!third_law && j < m_pdata->getN() && i > j)
+                continue;
+
+            pair_list->push_back(vec2(i, j));
+            }
+        }
+
+    unsigned long shape[] = {pair_list->size(), 2};
+    unsigned long stride[] = {2 * sizeof(uint32_t), sizeof(uint32_t)};
+
+    return pybind11::array_t(shape,                        // shape
+                             stride,                       // stride
+                             (uint32_t*)pair_list->data(), // data pointer
+                             delete_when_done);
+    }
+
+pybind11::object NeighborList::getPairListPython(uint64_t timestep)
+    {
+    compute(timestep);
+
+    bool third_law = getStorageMode() == NeighborList::half;
+
+#ifdef ENABLE_MPI
+    bool root = m_exec_conf->isRoot();
+#endif
+
+    ArrayHandle<unsigned int> h_n_neigh(getNNeighArray(), access_location::host, access_mode::read);
+    ArrayHandle<unsigned int> h_nlist(getNListArray(), access_location::host, access_mode::read);
+    ArrayHandle<size_t> h_head_list(getHeadList(), access_location::host, access_mode::read);
+
+    ArrayHandle<unsigned int> h_tags(m_pdata->getTags(), access_location::host, access_mode::read);
+
+    auto* pair_list = new std::vector<vec2<uint32_t>>();
+
+    for (int i = 0; i < (int)m_pdata->getN(); i++)
```

Comment:
Would be better to do:
```
for (unsigned int i = 0; i < m_pdata->getN(); i++)
```

Reply:
Also, I just noticed several usages of signed integers for offsets, number of elements... These should be `unsigned int` when used for particle indices or types and `size_t` when used for pair indices or counts.

Reply:
Most can definitely be changed to `unsigned int` and `size_t`. I'd also agree that the offsets and others should be `size_t`, but annoyingly many of the inputs to [`GatherV`](https://www.mpich.org/static/docs/v3.1/www3/MPI_Gatherv.html) require `int` or `int*` types, so things like `offsets` and `displacements` need to contain `int`s. I'm not sure if there is a way around it.

Reply:
Well, that is unfortunate. I guess we will need to store those in `int`s. Detect when there is an overflow and raise an exception. Users with simulations that have very large pair lists will need to use the local pair list API.

Reply:
MPI can be compiled with 64-bit addressing, but this seems to be rarely packaged. This has to be addressed somehow since this breaks #1173, but it seems like a big dependency upgrade to require all users build MPI from scratch. Maybe we could update the docs that a 64-bit MPI build is required for correct `pair_list` behavior under domain decomposition in large systems and include a helpful error message ...
> The `pair_list` is too large to be constructed with a 32-bit build of MPI.
> Please rebuild HOOMD with MPI and 64-bit addressing.

when when the neighborlist size is greater than `2^31 + 1`?

Reply:
Does MPI provide a preprocessor `#define` that you can use to tell when it is built for 64-bit? You would need that to conditionally compile in support for this code path.

This doesn't break the fix in #1173, as the simulations will still run correctly. It just means that users with excessively large pair lists will not be able to use the new `pair_list` member. This is not a big issue as the number of users who run simulations with more than 4 billion neighbors is rather small.

Reply:
Yes yes, it really is a very specific corner case. I'm pretty sure I can manage this with some preprocessor hacks.